### ISSUE TITLE
[prefer-readonly] False positive with Object.assign

### ISSUE THREAD

**Issue by boris-petrov** [typescript-eslint/typescript-eslint#2854]
**Repro**

```typescript
class X {
  private foo = 1;

  public setFoo(): void {
    Object.assign(this, { foo: 2 });
  }
}
```

This gives a `Member 'foo' is never reassigned; mark it as ``readonly``. eslint(@typescript-eslint/prefer-readonly)` warning which is wrong - I set it via `Object.assign`.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.9.0` |
| `@typescript-eslint/parser`        | `4.9.0` |
| `TypeScript`                       | `4.1.2` |
| `ESLint`                           | `7.15.0` |
| `node`                             | `15.3.0` |

**Comment by bradzacher** [comment]
There is no reasonable way for us to understand this.

TypeScript's own unused private member tooling does not understand this.

If you're using `Object.assign` in your classes, you're in a dynamic world where static analysis just cannot help you.
